### ISSUE TITLE
Check during Dataset Publish and Dataverse Create for Spam Content

### ISSUE THREAD

**Issue by djbrooke** [IQSS/dataverse#7291]
In #7238 I mentioned that we were seeing spam in Harvard Dataverse (dataverses and datasets) and I wanted to investigate putting in some captchas. It appears that the spammers aren't using automated means to create spam content but are instead doing it manually. This leads me to believe that a captcha or math problem solution would not be as effective as I'd like. :/ So, some new thoughts:

- We'd like to catch the spam before it's published instead of letting it be published and cleaning it up (@mheppler pointed out that letting users save something that couldn't be published doesn't really make sense, so if we can add a check on save that would make good sense)
- We have a workflow for locking datasets during publish for file integrity checks and PID registration, so if we wanted to run some publish-time check for spam we could incorporate it there and include it in the locking messaging.
- Our friends at Zenodo were nice enough to share a link to the repo that they use to identify spam at https://github.com/zenodo/zenodo-classifier. They are willing to accept contributions, and it may be good to collaborate on this because someone spamming Dataverse today is probably spamming Zenodo tomorrow, and vice versa. 
- How would we handle positives? Delete/Destroy? Keep them locked until a human can intervene? Move them to a holding Dataverse and delete after 7 days? Other? 🤔 

cc: @mheppler @landreev

**Comment by landreev** [comment]
What to do when you get a positive spam test result - not sure, we'll need to think about it some more. Maybe if it's a test-on-save, we just refuse to save, with a GUI error (similar to our normal validation errors). If it's on publish, we lock the dataset, with a special lock? 

Automatically deleting datasets is kind of a nuclear option. We can probably do that if we are absolutely confident - for example, if they entered a url of a known spam site. 

But we probably want to check for more subtle things as well; and then it becomes less straightforward. If we want to use some third party software for this, they don't usually give you a positive vs. negative test result, but a score; a likelihood of the message being spam. (Although I don't know if that's the case with the zenodo tool above; couldn't get that from the readme in their repo right away). But if you run a check and it tells us that there's a 60% chance this is spam - we probably don't want to delete it?

**Comment by djbrooke** [comment]
Thanks tech team for discussing today - my goal is to a. stop this stuff from getting into prod b. saving curation team time from manual clean up. My notes:

- @landreev was thinking of an external job to delete spam, on a cron (this would potentially not catch stuff before publication but could be run very often)
- @scolapasta brought up using the pre-publication workflows, this would result in the datasets/dataverses not getting published. this could use an externally running application
- We could even add in a check in the validation for the metadata entry - entered URLs vs. list of spam services

**Comment by landreev** [comment]
A status update: 
Much of the discussion above was operating under the assumption that the spam we've been seeing in production was mostly the same. Along the lines of largely the same copy-and-pasted text, urls pointing to the same few sites, etc. Having looked at the metadata collected by the curation team, this is simply not the case. There is an unexpected amount of variety; the URLs are rarely repeating (they may be redirected to a smaller number of repeating destinations - but they are constantly changing up the links, via different numeric ips, url shortening services etc.); the users who keep creating the dataverses and datasets are not coming from the same blocks of ip addresses. 

The earlier idea of an outside script was only based on such an assumption - that there was a very straightforward test for identifying this spam. There is no such simple, unambiguous test. And since creating a working and reliable test is going to take some time, we may as well add this as a filter to the application. 

I still don't think we want to add anything complicated to the application. For example, I'm not sure we need a dedicated workflow for this. I'm experimenting with just adding extra (optional) validation steps to the 2 publish commands. Under this setup an installation will be able to define an external executable; if specified, the publish(Dataverse|Dataset) command will run it on the current object metadata and check the exit code. If the test comes back positive, it will simply refuse to publish (I'm not bothering with any dedicated locks at this point), sending a notification to the author ("we cannot publish your ... since our test has indicated that it may be spam. if you think this is an error, please contact ...") and the admins. I'll make a PR for this; if we want to expand this model - via workflows or whatever, somebody else could pick it up. 

But, again, the application-level mods are not the difficult part. The actual test is. I've been mostly experimenting with SpamAssasin. (And a 3rd party module Text::SpamAssasin, that allows to run SA on arbitrary chunks of text/html, instead of emails). It's promising, but it is not by any means straightforward. It's not a test that just works out of the box. It can only work once we create  enough "rules" for it to recognize the spam we've been seeing. (And it will almost certainly be an ongoing arms race, with the spammers refining their craft, and us having to counter-act with better rules). Some individual rules can be rather straightforward. For example, it appears to be safe to assign a spam point (or even two) for URLs with numeric ips in them. Others can be more controversial, but still effective (it appears that detecting a certain language can be worth a couple of points; this will occasionally ring true for a legitimate dataset from that region, but if the cutoff is high enough - say 5 points - the language alone would not trigger a false positive...)
Basically, you get the idea - it's looking like work, no matter how you slice it.

**Comment by scolapasta** [comment]
I still think it will make sense to use the workflow model for this; I know it means making some changes to that, but I believe those changes would be useful there, outside of just spam, and would help enable workflows to be used for other things (Data curation, maybe?). But I also get that it's more work.

Regardless of which approach, once thing we should consider is not calling this check (via the code or the workflow) for superusers. That way, we have an easy way to override the false positive once they've contacted us.

**Comment by landreev** [comment]
Yes, I thought of that part - that superuser should be able to override the check.

**Comment by landreev** [comment]
> ... But I also get that it's more work.

It is, but it would still be less work than creating a reliable detection test; I'm pretty sure. So once we have that, we can definitely experiment with how much we want to change in the application.

**Comment by qqmyers** [comment]
FWIW: I think creating a workflow step that executes a local code and returns the status (or success/fail based on that) would be pretty straight forward. Transitioning from there to using the web 'wait for response' step could then allow you to get the detector off the DV machine, etc. Configuring as a pre-publish workflow would trigger all the normal workflow locking/unlocking so there shouldn't be extra work to handle that. I'd have to check to make sure that everything that's needed can be sent to a step, but I think that's true (otherwise the archiving workflows wouldn't work).

**Comment by landreev** [comment]
As I mentioned earlier, I've done relatively little work on the dataverse side as part of this. Having mostly focused on developing actual external tests capable of reliably detecting spam. This is all I'm doing on the Dataverse side so far, in PublishDataverse and  PublishDataset commands: if an external validation script is defined for either of the two, the command runs it on the metadata, and throws a command exception if the test fails. So it simply refuses to publish, leaving the dataset in draft or the dataverse unpublished. There are no special locks; and no attempts to delete the offending dataverse or dataset. 

Since @scolapasta et al have suggested expanding this dataverse-side handling of the checks - perhaps by introducing a dedicated workflow for dataset publishing etc., I have checked in all my current dataverse-side mods in case anyone wants to pick that part up. (I can add an a quick note on how to define a mock failing test to be able to work on this).

**Comment by landreev** [comment]
The above can be tested/worked on without having a real spam check in place. 
For example, the following will specify a dataverse metadata check that will always fail (by configuring `/usr/bin/false` as the executable to run):
```
curl -X PUT -d '/usr/bin/false' http://localhost:8080/api/admin/settings/:DataverseMetadataValidatorScript
```
The setting for datasets is `:DatasetMetadataValidatorScript`.

**Comment by landreev** [comment]
I checked in a SpamAssassin-based  solution for dataverses (a dataset solution is largely identical - the difference being more text fields to extract and check from the json metadata export). 
Other changes on the application side made last week include a configurable rejection message. 
For example: 
```
curl -X PUT -d 'This looks like spam! But nice try.' http://localhost:8080/api/admin/settings/:DataverseMetadataValidationFailureMsg
```

![Screen Shot 2020-12-08 at 9 36 31 AM](https://user-images.githubusercontent.com/737558/101497956-890a3a00-3939-11eb-988d-c6f5e447a251.png)

Spam check override can be enabled for the admins: 

```
curl -X PUT -d true http://localhost:8080/api/admin/settings/:ExternalValidationAdminOverride
```

**Comment by djbrooke** [comment]
I'm going to close this issue. Thanks @landreev and others for the work here, and double thanks for getting it deployed on dataverse.harvard.edu before the holiday break so that we didn't need to fend off spam manually during the time off.

I've added the details for our implementation on the Harvard Dataverse Repository to a folder in the dataverse-security repo so that we can provide this to any other installations that encounter issues with spam. I'm hopeful that other installations won't be as impacted by this, because either they are not as high profile as the Harvard Dataverse Repository and/or their workflows are more locked down in regards to who can create Dataverse collections and datasets. 

We are happy enough with this solution on the Harvard Dataverse Repository that I don't want to spend additional time on this, but we can revisit if necessary. These will be specific issues that we create, prioritize, and take on.
